### DISCUSSION TITLE
How can I create a custom task including environment, scene, robot, and objects?

### DISCUSSION THREAD

**Discussion by RZFan525** [StanfordVL/BEHAVIOR-1K#972] [Q&A] [answered]
I want to create a PUT task which is aimed to PUT something on something, such as put the apple into the plate. And, this environment contains a desk, an apple, a plate, and a franka robot. And I also want to random the position of the apple and the plate in range of the desk and the distance of robot. So I cannot assign the position by exact x, y, z in config.

I have read the doc and know how to set config and create a environment. But I don't know how to random the position in an exact range. And I also don't know the relation between environment and task in your doc. I mean the success detection function in tasks should use the object name and position. However, the object information is in environment. which cannot be known by task class.

Is there any example to create a task like I describe above?

Thank you!

**Comment by hang-yin** [comment, marked as answer]
Hi @RZFan525 , thanks for reaching out!

Here's my understanding of your problem: 
1. You want a Franka, an apple, and a plate to be on top of a desk
2. The apple and the plate need to be within the robot's workspace

Here's a workflow that should work: 
1. Create custom task definition
  a. We will need to first uninstall the pip-installed BDDL: `pip uninstall bddl` in your conda environment
  b. Install BDDL from source so that we can add a new task: `git clone https://github.com/StanfordVL/bddl.git`
  c. Install BDDL: `cd bddl` & `pip install -e .`
  d. Write a custom task definition - under `bddl/activity_definitions`, create a new directory called `demo_put_task`. Within that directory, create a new file `problem0.bddl` with the following content: 
```
(define (problem demo_put_task-0)
    (:domain omnigibson)

    (:objects
        breakfast_table.n.01_1 - breakfast_table.n.01
        apple.n.01_1 - apple.n.01
        plate.n.04_1 - plate.n.04
        floor.n.01_1 - floor.n.01
        agent.n.01_1 - agent.n.01
    )
    
    (:init 
        (ontop plate.n.04_1 breakfast_table.n.01_1)
        (ontop apple.n.01_1 breakfast_table.n.01_1) 
        (inroom breakfast_table.n.01_1 living_room)
        (inroom floor.n.01_1 living_room)
    )
    
    (:goal 
        (and 
            (ontop apple.n.01_1 plate.n.04_1)
        )
    )
)
```
2. Usually, BEHAVIOR tasks are attempted with Mobile Manipulators. However, Franka is a fixed base manipulation robot without a mobile base, which means we can not directly sample it. Here's what you need to do to get this working - in `omnigibson/utils/bddl_utils.py:BDDLSampler`, there's a `_build_sampling_order` method. This method has this segment of code:
```
        # Possibly remove the agent entity if we're in an empty scene -- i.e.: no kinematic sampling needed for the
        # agent
        if self._scene_model is None:
            remaining_kinematic_entities -= {"agent.n.01_1"}
```
In our case, we will need to always remove this agent. You can simply remove this if check to always do this. 
3. Finally, we can write a script to do rejection sampling so that the apple and plate are always in range of the Franka: 
```
import omnigibson as og
from omnigibson.object_states import *
from omnigibson.utils.object_state_utils import sample_kinematics
from omnigibson.action_primitives.starter_semantic_action_primitives import StarterSemanticActionPrimitives

cfg = {
    "scene": {
        "type": "InteractiveTraversableScene",
        "scene_model": "Rs_int",
        "load_object_categories": ["floors", "breakfast_table"],
    },
    "robots": [
        {
            "type": "FrankaPanda",
            "obs_modalities": ["rgb"],
        },
    ],
    "task": {
        "type": "BehaviorTask",
        "activity_name": "demo_put_task",
        "activity_definition_id": 0,
        "activity_instance_id": 0,
        "online_object_sampling": True,
    },
}

valid_sample = False

while not valid_sample:
    env = og.Environment(configs=cfg)
    robot = env.scene.robots[0]

    tables = []
    for obj in env.scene.objects:
        if obj.category == "breakfast_table":
            tables.append(obj)
        elif obj.category == "apple":
            apple = obj
        elif obj.category == "plate":
            plate = obj

    sampled_table = None
    for table in tables:
        if apple.states[OnTop].get_value(table) and plate.states[OnTop].get_value(table):
            # This is the table we are looking for
            sampled_table = table
            break
    assert sampled_table is not None, "Could not find a table for the apple and plate to be on top of"

    robot.reset()
    sample_kinematics("onTop", robot, sampled_table)

    ssap = StarterSemanticActionPrimitives(env, enable_head_tracking=False)
    apple_in_reach = ssap._target_in_reach_of_robot(apple.get_position_orientation())
    plate_in_reach = ssap._target_in_reach_of_robot(plate.get_position_orientation())
    valid_sample = apple_in_reach and plate_in_reach
    if not valid_sample:
        og.clear()

valid_env = env
breakpoint()
# Do stuff ........

```

Let me know if this works for you~!

**Comment by RZFan525** [comment]
Thank you very much!!!! It's very helpful for me!!! I understand it. 
And, can I modify the BBDL of scene of office_vendor_machine in InteractiveTraversableScene class to make my put task into this existing environment?  I mean I want to put the apple, franka, and plate on the conference table of office_vendor_machine scene.

**Comment by RZFan525** [comment]
There is an error about the states:
```
if apple.states['OnTop'].get_value(table) and plate.states['OnTop'].get_value(table):
KeyError: 'OnTop'
```

**Reply by RZFan525** [reply]
> There is an error about the states:
> 
> ```
> if apple.states['OnTop'].get_value(table) and plate.states['OnTop'].get_value(table):
> KeyError: 'OnTop'
> ```

I'm sorry. I delete the quotation mark and it's ok now.

**Comment by RZFan525** [comment]
I see the BEHAVIOR1k scene is writen by json. I want to import a BEHAVIOR1k scene and add an apple, a franka and a plate in their table. Can I finish it by BDDL?

**Comment by hang-yin** [comment]
Hi @RZFan525! Glad it worked! 
I have just recently started learning about BDDL and how it interacts with our stack, but here's my understanding:

- If the object you want to use (e.g. conference table) is already in the scene you are using (e.g. `office_vendor_machine`), when writing the BDDL task definition, you should use the `inroom` keyword. For example, in my script, I wanted to use the breakfast table in `Rs_Int`, so I wrote something like `(inroom breakfast_table.n.01_1 living_room)`. To get the specific category of the object you are looking for (e.g. breakfast_table), you can use our `omnigibson/examples/scenes/scene_tour_demo.py` to inspect everything in a scene. In your case, you might want to do something like:
```
        (inroom conference_table.n.01_1 meeting_room) 
        (inroom floor.n.01_1 meeting_room) 
        (ontop agent.n.01_1 floor.n.01_1)
```
 - If the object you are looking for is not in the scene (e.g. apple and plate), you need to specify its relationship to a scene object (e.g. conference_table.n.01_1). During task sampling, they will be randomly sampled in relation to that scene object. 

I hope this helps! Let me know if you run into any other problems.

**Reply by RZFan525** [reply]
> Hi @RZFan525! Glad it worked! I have just recently started learning about BDDL and how it interacts with our stack, but here's my understanding:
> 
>     * If the object you want to use (e.g. conference table) is already in the scene you are using (e.g. `office_vendor_machine`), when writing the BDDL task definition, you should use the `inroom` keyword. For example, in my script, I wanted to use the breakfast table in `Rs_Int`, so I wrote something like `(inroom breakfast_table.n.01_1 living_room)`. To get the specific category of the object you are looking for (e.g. breakfast_table), you can use our `omnigibson/examples/scenes/scene_tour_demo.py` to inspect everything in a scene. In your case, you might want to do something like:
> 
> 
> ```
>         (inroom conference_table.n.01_1 meeting_room) 
>         (inroom floor.n.01_1 meeting_room) 
>         (ontop agent.n.01_1 floor.n.01_1)
> ```
> 
>     * If the object you are looking for is not in the scene (e.g. apple and plate), you need to specify its relationship to a scene object (e.g. conference_table.n.01_1). During task sampling, they will be randomly sampled in relation to that scene object.
> 
> 
> I hope this helps! Let me know if you run into any other problems.

I have tested this BDDL. However, the apple and the plate are not in the room. My BDDL file is below:
```
(define (problem demo_put_task-0)
    (:domain omnigibson)

    (:objects
        conference_table.n.01_1 - conference_table.n.01
        apple.n.01_1 - apple.n.01
        plate.n.04_1 - plate.n.04
        floor.n.01_1 - floor.n.01
        agent.n.01_1 - agent.n.01
    )
    
    (:init 
        (ontop plate.n.04_1 conference_table.n.01_1)
        (ontop apple.n.01_1 conference_table.n.01_1) 
        (inroom conference_table.n.01_1 meeting_room)
        (inroom floor.n.01_1 meeting_room)
    )
    
    (:goal 
        (and 
            (ontop apple.n.01_1 plate.n.04_1)
        )
    )
)
```

**Comment by RZFan525** [comment]
Thank you !!! 
I'm wondering how to know the parameters (e.g., meeting_room, living_room) and what is the xxx.n.01_1 (What is the number meaning?). How to connect parameters and the objects in the environment.

**Comment by hang-yin** [comment]
You are welcome @RZFan525 ! 

To get each of the room name, you can take a look at our scene knowledgebase: https://behavior.stanford.edu/knowledgebase/scenes/index.html

When you click on the scene you are looking for, it should show you all the rooms available and what objects are present in each of them. 

As of the `xxx.n.01_1`, this is the synset of an object plus its instance #. If you go to our object knowledgebase https://behavior.stanford.edu/knowledgebase/objects/index.html, you can see that each object is associated with a synset. 

For example, `conference_table.n.01_1` means that we are referring to the `conference_table.n.01` synset, and this is the number 1 object under this synset. 

If you have multiple objects under the same synset, you can do something like: 
`apple.n.01_1 apple.n.01_2 apple.n.01_3 - apple.n.01`
